wgengine/tstun/tun.go

```diff
@@ -61,10 +66,26 @@ type TUN struct {
 	// the other direction must wait on a Wireguard goroutine to poll it.
 	outbound chan []byte
 
-	// fitler stores the currently active package filter
+	// filter is the main packet filter set based on common rules and ACLs.
+	// Unlike pre-/post-filters, it is updated at runtime by the control client.
 	filter atomic.Value // of *filter.Filter
 	// filterFlags control the verbosity of logging packet drops/accepts.
 	filterFlags filter.RunFlags
+
+	// The following are exported, but not synchronized in any way.
+	// The intent is for them to be initialized once and not touched afterward.
+
+	// PreFilterIn are the inbound filter functions that run before the main filter
+	// and therefore see the packets that are later dropped by it.
+	PreFilterIn []FilterFunc
```

Comment:
What is the motivation for making these slices, i.e. what is an example of more than one function here?

Especially given the comment above that these should be initialized once, that means in one place, it seems like it would be easy enough for the initializer to combine some set of functions into a single function at the initialization site.

Reply:
Very fair, I'll change that. Relatedly, not having these as concrete functions moves `q` to the heap, which is unfortunate. Not sure if having this extensibility is valuable enough to justify working around that (a `sync.Pool` of `q.Decode`s?)

Reply:
Solved with `sync.Pool`